File: visit-scheduler-web/src/schema/schedule.ts
import { z } from "zod";

import { dateSchema } from "@/lib/datetime";
import { naturalNumberSchema } from "@/schema/common";
import { comparisonOperatorSchema } from "@/schema/enum";
import { chouseisanIdSchema, scheduleGroupIdSchema, visitUserIdSchema } from "@/schema/id";

////////////////////// Schedule Group ////////////////////////

export const scheduleGroupSchema = z.object({
  id: scheduleGroupIdSchema,
  name: z.string(),
  tasks: z.array(z.string()),
  description: z.string().optional(),
});

export type ScheduleGroup = z.infer<typeof scheduleGroupSchema>;

////////////////////// Schedule Master ////////////////////////

export const scheduleMasterSchema = z.object({
  chouseisan_id: chouseisanIdSchema,
  title: z.string(),
  candidates: z.array(dateSchema),
  merge_with: z.array(chouseisanIdSchema),
});

export type ScheduleMaster = z.infer<typeof scheduleMasterSchema>;

////////////////////// Sync Chouseisan Result ////////////////////////

export const syncChouseisanResultSchema = z.object({
  schedule_master: scheduleMasterSchema,
  diff_visit_users: z.array(z.string()),
});

export type SyncChouseisanResult = z.infer<typeof syncChouseisanResultSchema>;

////////////////////// Optimize Config ////////////////////////

export const entryCohortConstraintSchema = z.array(
  z.tuple([z.array(naturalNumberSchema), comparisonOperatorSchema, naturalNumberSchema]),
);

export type EntryCohortConstraint = z.infer<typeof entryCohortConstraintSchema>;

export const constraintsByCandidateSchema = z.object({
  candidate: dateSchema,
  group: z
    .string()
    .nullable()
    .optional()
    .transform((r) => r ?? null),
  tasks: z.array(z.tuple([z.string(), z.number().int()])),
  irregular_member_count_constraint: naturalNumberSchema
    .nullable()
    .optional()
    .transform((r) => r ?? null),
});

export type ConstraintsByCandidate = z.infer<typeof constraintsByCandidateSchema>;

export const optimizeConfigSchema = z.object({
  chouseisan_id: chouseisanIdSchema,
  entry_cohort_constraint: entryCohortConstraintSchema,
  driver_level_constraint: z.number().int().nonnegative(),
  basic_member_count_constraint: naturalNumberSchema,
  gender_consideration: z.array(naturalNumberSchema),
  candidate_constraints: z.array(constraintsByCandidateSchema),
});

export type OptimizeConfig = z.infer<typeof optimizeConfigSchema>;

////////////////////// Optimized Schedule ////////////////////////

export const pickedDateWithMembersSchema = z.object({
  date: dateSchema,
  member: z.array(visitUserIdSchema),
});

export type PickedDateWithMembers = z.infer<typeof pickedDateWithMembersSchema>;

export const optimizedScheduleSchema = z.object({
  chouseisan_id: chouseisanIdSchema,
  schedule: z.array(pickedDateWithMembersSchema),
});

export type OptimizedSchedule = z.infer<typeof optimizedScheduleSchema>;

////////////////////// Schedule With Config ////////////////////////

export const scheduleWithConfigSchema = z.object({
  schedule_master: scheduleMasterSchema,
  optimize_config: optimizeConfigSchema
    .optional()
    .nullable()
    .transform((r) => r ?? null),
  optimized_schedule: optimizedScheduleSchema
    .optional()
    .nullable()
    .transform((r) => r ?? null),
});

export type ScheduleWithConfig = z.infer<typeof scheduleWithConfigSchema>;

////////////////////// Optimization Result ////////////////////////

export const optimizationResultSchema = z
  .object({
    chouseisan_id: chouseisanIdSchema,
    status: z.literal("optimal"),
    optimized_schedule: optimizedScheduleSchema,
  })
  .or(
    z.object({
      chouseisan_id: chouseisanIdSchema,
      status: z.literal("infeasible").or(z.literal("failed")),
      optimized_schedule: z.null(),
    }),
  );

export type OptimizationResult = z.infer<typeof optimizationResultSchema>;

File: visit-scheduler-api/src/schema/schedule.py
from datetime import date, datetime
from typing import List, Optional, Dict

from lib.datetime import date_to_datetime
from schema.config import VSBaseModel
from schema.enum import VSOptimizationStatus
from src.schema.optimize_config import OptimizeConfig


class ScheduleMaster(VSBaseModel):
    chouseisan_id: str
    title: str
    candidates: list[date]
    merge_with: list[str]

    def to_dto(self) -> "ScheduleMasterDto":
        candidates = [date_to_datetime(candidate) for candidate in self.candidates]

        return ScheduleMasterDto(
            chouseisan_id=self.chouseisan_id,
            title=self.title,
            candidates=candidates,
            merge_with=self.merge_with,
        )


# NOTE: firesotore
#   firestore に保存できる構造
#   - date は datetime に変換して保存する
#   - タスク一覧は別で更新する必要がある
class ScheduleMasterDto(VSBaseModel):
    chouseisan_id: str
    title: str
    candidates: list[datetime]
    merge_with: Optional[list[str]] = None

    def to_model(self) -> ScheduleMaster:
        candidates = [candidate.date() for candidate in self.candidates]
        merge_with = self.merge_with
        if merge_with is None:
            merge_with = []

        return ScheduleMaster(
            chouseisan_id=self.chouseisan_id,
            title=self.title,
            candidates=candidates,
            merge_with=merge_with,
        )


class SyncChouseisanResult(VSBaseModel):
    schedule_master: ScheduleMaster
    diff_visit_users: List[str]


class PickedDateWithMember(VSBaseModel):
    date: date
    member: List[str]  # visit_user_id


class OptimizedSchedule(VSBaseModel):
    chouseisan_id: str
    schedule: List[PickedDateWithMember]

    def to_dto(self) -> "OptimizedScheduleDto":
        schedule = {}
        for date_with_member in self.schedule:
            str_date = date_with_member.date.strftime("%Y-%m-%d")
            schedule[str_date] = date_with_member.member

        return OptimizedScheduleDto(chouseisan_id=self.chouseisan_id, schedule=schedule)


class OptimizedScheduleDto(VSBaseModel):
    chouseisan_id: str
    schedule: Dict[str, List[str]]

    def to_model(self) -> OptimizedSchedule:
        schedule = []
        for str_date in self.schedule:
            date = datetime.strptime(str_date, "%Y-%m-%d").date()
            schedule.append(
                PickedDateWithMember(date=date, member=self.schedule[str_date])
            )

        return OptimizedSchedule(chouseisan_id=self.chouseisan_id, schedule=schedule)


class ScheduleWithConfig(VSBaseModel):
    schedule_master: ScheduleMaster
    optimize_config: Optional[OptimizeConfig]
    optimized_schedule: Optional[OptimizedSchedule]


class OptimizationResult(VSBaseModel):
    chouseisan_id: str
    status: VSOptimizationStatus
    optimized_schedule: Optional[OptimizedSchedule]
